rfcs/2020-04-04-2221-kubernetes-integration.md

```diff
@@ -0,0 +1,253 @@
+
+### Tutorial
+
+#### Kubectl Interface
+
+1.  Configure Vector:
+
+    Before we can deplo Vector we must configure. This is done by creating
+    a Kubernetes `ConfigMap`:
+
+    ...insert selector to select any of Vector's sinks...
+
+    ```bash
+    echo '
+    apiVersion: v1
+    kind: ConfigMap
+    metadata:
+      name: vector-config
+      namespace: logging
+      labels:
+        k8s-app: vector
+    data:
+      vector.toml: |
+        # Docs: https://vector.dev/docs/
+
+        # Set global options
+        data_dir = "/var/tmp/vector"
+
+        # Ingest logs from Kubernetes
+        [sources.kubernetes]
+          type = "kubernetes"
+
+        # Enrich logs with Pod metadata
+        [transforms.pod_metadata]
+          type = "kubernetes_pod_metadata"
+          inputs = ["kubernetes"]
+
+        # Send data to one or more sinks!
+        [sinks.aws_s3]
+          type = "aws_s3"
+          inputs = ["pod_metadata"]
+          bucket = "my-bucket"
+          compression = "gzip"
+          region = "us-east-1"
+          key_prefix = "date=%F/"
+    ' > vector-configmap.toml
+    ```
+
+2.  Deploy Vector!
+
+    Now that you have your custom `ConfigMap` ready it's time to deploy
+    Vector. To ensure Vector is isolated and has the necessary permissions
+    we must create a `namespace`, `ServiceAccount`, `ClusterRole`, and
+    `ClusterRoleBinding`:
+
+    ```bash
+    kubectl create namespace logging
+    kubectl create -f vector-service-account.yaml
+    kubectl create -f vector-role.yaml
+    kubectl create -f vector-role-binding.yaml
+    kubectl create -f vector-configmap.yaml
+    kubectl create -f vector-daemonset.yaml
+    ```
+
+    * *See [outstanding questions 3, 4, 5, 6, and 7](#outstanding-questions).*
+
+    That's it!
+
+#### Helm Interface
+
+TODO: fill in
+
+## Prior Art
+
+1. [Filebeat k8s integration]
+1. [Fluentbit k8s integration]
+2. [Fluentd k8s integration]
+3. [LogDNA k8s integration]
+4. [Honeycomb integration]
+3. [Bonzai logging operator] - This is approach is likely outside of the scope
+   of Vector's initial Kubernetes integration because it focuses more on
+   deployment strategies and topologies. There are likely some very useful
+   and interesting tactics in their approach though.
+4. [Influx Helm charts]
+
+## Sales Pitch
+
+See [motivation](#motivation).
+
+## Drawbacks
+
+1. Increases the surface area that our team must manage.
+
+## Alternatives
+
+1. Not do this integration and rely solely on external community driven
+   integrations.
+
+## Outstanding Questions
+
+1. What is the minimal Kubernetes version that we want to support. See
+   [this comment][kubernetes_version_comment].
+1. What is the best to avoid Vector from ingesting it's own logs? I'm assuming
+   that my [`kubectl` tutoria](#kubectl-interface) handles this with namespaces?
+   We'd just need to configure Vector to excluse this namespace?
+1. I've seen two different installation strategies. For example, Fluentd offers
```

Comment:
I'd use a minimal example in the documentation (thus a single daemonset and configmap) and redirect to helm chart for full setup.

Reply:
I'd go with
- a single YAML file with all the necessary stuff (i.e. `DeamonSet`, `Role`, `RoleBinding`, etc) that we ship, and that is namespaced
- maybe another YAML file for stuff the we ship and that's not namespaced (`ClusterRole`, `ClusterRoleBinding`)
- a user-provided file with `ConfigMap` and everything else that we want users to provide.

Thus, either two or three files.

I'll update the RFC accordingly.